File: src/main/java/no/nav/foreldrepenger/stønadskonto/regelmodell/StønadskontoRegelOrkestrering.java
package no.nav.foreldrepenger.stønadskonto.regelmodell;

import no.nav.foreldrepenger.stønadskonto.regelmodell.grunnlag.BeregnKontoerGrunnlag;
import no.nav.foreldrepenger.stønadskonto.regelmodell.regler.BeregnKontoer;
import no.nav.foreldrepenger.stønadskonto.regelmodell.regler.KontoerMellomregning;

public class StønadskontoRegelOrkestrering {

    public StønadskontoResultat beregnKontoer(BeregnKontoerGrunnlag grunnlag) {
        if (grunnlag.getDekningsgrad() == null) {
            throw new IllegalArgumentException("Mangler dekningsgrad");
        }
        var mellomregning = new KontoerMellomregning(grunnlag);
        var beregnKontoer = new BeregnKontoer();
        beregnKontoer.evaluer(mellomregning);

        return new StønadskontoResultat(mellomregning.getBeregnet());
    }
}

File: stonadskonto/src/StønadskontoRegelOrkestrering.ts
import { BeregnKontoerGrunnlag } from './grunnlag/BeregnKontoerGrunnlag';
import { StønadskontoResultat } from './StønadskontoResultat';
import { KontoerMellomregning } from './regler/KontoerMellomregning';
import {
  fastsettStønadskontoStruktur,
  sjekkOmFødsel,
  leggTilDagerVedFødsel,
  sjekkOmMerEnnEttBarn,
  leggTilDagerVedFlereBarn,
  sjekkOmBareFarHarRett,
  leggTilDagerVedBareFarRett,
  sjekkOmTetteSaker,
  leggTilMinsterettVedTetteSaker,
  opprettKontoer,
  ferdigBeregnetKontoer,
} from './regler/Rules';

export class StønadskontoRegelOrkestrering {
  beregnKontoer(grunnlag: BeregnKontoerGrunnlag): StønadskontoResultat {
    if (!grunnlag.getDekningsgrad()) {
      throw new Error('Mangler dekningsgrad');
    }

    const mellomregning = new KontoerMellomregning(grunnlag);

    fastsettStønadskontoStruktur(mellomregning);

    if (sjekkOmFødsel(mellomregning)) {
      leggTilDagerVedFødsel(mellomregning);
    }

    if (sjekkOmMerEnnEttBarn(mellomregning)) {
      leggTilDagerVedFlereBarn(mellomregning);
    }

    if (sjekkOmBareFarHarRett(mellomregning)) {
      leggTilDagerVedBareFarRett(mellomregning);
    }

    if (sjekkOmTetteSaker(mellomregning)) {
      leggTilMinsterettVedTetteSaker(mellomregning);
    }

    opprettKontoer(mellomregning);
    ferdigBeregnetKontoer(mellomregning);

    return new StønadskontoResultat(mellomregning.getBeregnet());
  }
}

